docs: Document the new LoadBalancer deployment option (#1350)
redpanda-data/documentation: docs/manage/kubernetes/networking/configure-external-access-loadbalancer.mdx

<head>
    <meta name="title" content="Configure External Access through LoadBalancer Services | Redpanda Docs"/>
    <meta name="description" content="Expose your Redpanda cluster to clients outside of your Kubernetes cluster by using LoadBalancer Services."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ValuesYaml from '../shared/_values-yaml.mdx'

Learn how to expose your Redpanda cluster to clients outside of your Kubernetes cluster by using LoadBalancer Services.

Here's what you'll need to do:

1. Set the `external.type` field to `LoadBalancer` to create a LoadBalancer Service for each replica in the StatefulSet.
1. Advertise your brokers to the outside world using either a [custom domain](#advertise-a-custom-domain) or the [DNS names or IP addresses of their load balancer instances](#advertise-the-default-addresses-of-the-load-balancer-instances).

:::info

> For clusters with TLS enabled, advertising the DNS names or IP addresses of load balancer instances is not supported. Instead, use [custom domains](#advertise-a-custom-domain).

Reviewer: Redpanda has TLS enabled by default? If so, clarify here to expect to use custom domains.

:::

## How external access works with LoadBalancer Services

When the `external.type` field is set to `LoadBalancer`, the Helm chart creates one LoadBalancer Service for each replica in the StatefulSet. Each LoadBalancer Service routes external traffic to the listener ports on a single Redpanda broker.

> The Helm chart adds the `spec.selector.statefulset.kubernetes.io/pod-name` setting to each LoadBalancer Service so that each one targets only one Redpanda broker. Each Pod in a StatefulSet is automatically given the `statefulset.kubernetes.io/pod-name` label, which contains the name of the Pod. By setting the `spec.selector.statefulset.kubernetes.io/pod-name` field of the LoadBalancer Service to the name of a Pod, it makes the LoadBalancer Service target only the individual Redpanda broker and not the whole cluster.

Reviewer: ```suggestion
The Helm chart adds the `spec.selector.statefulset.kubernetes.io/pod-name` setting to each LoadBalancer Service so that each targets only one Redpanda broker. Each Pod in a StatefulSet is automatically given the `statefulset.kubernetes.io/pod-name` label, which contains the name of the Pod. By setting the `spec.selector.statefulset.kubernetes.io/pod-name` field of the LoadBalancer Service to the name of a Pod, it makes the LoadBalancer Service target only the individual Redpanda broker and not the whole cluster.
```

import TabItem from '@theme/TabItem';
import ValuesYaml from '../shared/_values-yaml.mdx'

> When the `external.type` field is set to `LoadBalancer`, the Helm chart creates one LoadBalancer Service for each replica in the StatefulSet. Each LoadBalancer routes external traffic to the listener ports on a single Redpanda broker.

Reviewer: The updated content doesn't have a descriptive summary about the purpose of this topic that the removed sentence did.

```suggestion
Expose your Redpanda cluster to clients outside of your Kubernetes cluster by using LoadBalancer Services.

When the `external.type` field is set to `LoadBalancer`, the Helm chart creates one LoadBalancer Service for each replica in the StatefulSet. Each LoadBalancer routes external traffic to the listener ports on a single Redpanda broker.
```

The Helm chart adds the `spec.selector.statefulset.kubernetes.io/pod-name` setting to each LoadBalancer Service so that each one targets only one Redpanda broker. Each Pod in a StatefulSet is automatically given the `statefulset.kubernetes.io/pod-name` label, which contains the name of the Pod. By setting the `spec.selector.statefulset.kubernetes.io/pod-name` field of the LoadBalancer Service to the name of a Pod, it makes the LoadBalancer Service target only the individual Redpanda broker and not the whole cluster.

To expose your Redpanda cluster externally through a LoadBalancer Service, your brokers must advertise one of the following:

- [A custom domain](#advertise-a-custom-domain)
- [The DNS names or IP addresses of their load balancer instances](#advertise-the-default-addresses-of-the-load-balancer-instances) (not supported for clusters with TLS enabled)

## Prerequisites

> - One worker node for each Redpanda broker that you want to deploy.

Reviewer: `Run`, `create`, `use` one worker node?

```suggestion
- Run one worker node for each Redpanda broker that you want to deploy.
```

When the `external.type` field is set to `LoadBalancer`, the Helm chart creates one LoadBalancer Service for each replica in the StatefulSet. Each LoadBalancer routes external traffic to the listener ports on a single Redpanda broker.

> ![A client connects to a Pod directly through a load balancer](./loadbalancer.png)

Reviewer: Skimming `A client connects to a Pod directly` could be misunderstood as there being nothing between the external client and Pod. Removing `directly` prevents that misunderstanding and preserves the intended meaning:

  ```suggestion
![A client connects to a Pod through a load balancer](./loadbalancer.png)
```

  external:
    domain: <custom-domain>
      enabled: true

>       type: LoadBalancer

Author: If a user is running on a deployment that has external access disabled, they need to know to enable it for this doc to work. I think it's best to be explicit about what settings are required even if they are currently the defaults.

  Hostname| Load balancer
  |---|---
  `<subdomain-for-broker-0>`|`<load-balancer-address-0>`
  `<subdomain-for-broker-1>`|`<load-balancer-address-1>`
  `<subdomain-for-broker-2>`|`<load-balancer-address-2>`

1. Wait for your DNS changes to be propagated.

1. Use your custom domain to communicate with the Redpanda cluster from outside the Kubernetes cluster:

  <Tabs groupId="tls" queryString>
  <TabItem value="enabled" label="TLS Enabled">

    1. Save the root certificate authority (CA) to your local file system outside Kubernetes:

      ```bash
      kubectl -n redpanda get secret redpanda-default-root-certificate -o go-template='{{ index .data "ca.crt" | base64decode }}' > ca.crt
      ```

    1. Get information about the cluster:

      ```bash
      rpk cluster info \
        --tls-enabled \
        --tls-truststore=ca.crt \
        --brokers <subdomain-for-broker-0>.<custom-domain>
      ```

  </TabItem>
  <TabItem value="disabled" label="TLS Disabled">

    ```bash
    rpk cluster info --brokers <subdomain-for-broker-0>.<custom-domain>
    ```

  </TabItem>
  </Tabs>

## Advertise the default addresses of the load balancer instances

> You can configure each Redpanda broker to advertise the DNS name or IP address of its corresonding load balancer instance.

Reviewer 2: ```suggestion
You may not know your LoadBalancer services' addresses prior to deployment. In this case, you can configure each Redpanda broker to advertise the DNS name or IP address of its corresponding load balancer resource. This requires an initial deployment step followed by a deployment upgrade that updates the advertised listener.
```

:::info

> If your cluster has TLS enabled, you must [advertise custom domains](#advertise-custom-domains) because the Helm charts adds only custom domains to the SAN list of TLS certificates, not IP addresses.

Reviewer 2: ```suggestion
If your cluster has TLS enabled, you must [advertise custom domains](#advertise-custom-domains) because the helm chart only adds custom domains to the SAN list of TLS certificates (not IP addresses).

If your LoadBalancer services have IP addresses assigned rather than DNS names, then you must make those IP addresses resolvable by DNS names for secure TLS access. For development purposes, adding `/etc/hosts` entries will work. But for production you will need to update your organization's DNS service.
```

Reviewer 2: This makes it more clear that you can still use TLS if your LoadBalancer services have an IP (rather than a DNS name) assigned... you just need to make that IP address resolvable by the DNS address that works with your cert.

:::

> If you're hosting Redpanda on a managed Kubernetes platform, follow the steps in the **Managed** tab. Otherwise, follow the steps in the **Bare-Metal** tab.
> 
> <Tabs groupId="kubernetes-platforms">
> <TabItem value="managed" label="Managed" default>
> 

Reviewer 2: ```suggestion
```

1. Deploy Redpanda with TLS disabled and enable the LoadBalancer Service type:

  <Tabs groupId="helm-config">
  <TabItem value="values" label="--values">

  ```yaml title="loadbalancer-tls-disabled.yaml"
  external:

>     enabled: true

Reviewer 2: ```suggestion
```

    type: LoadBalancer
  tls:
    enabled: false
  ```

  ```bash

>   helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
>     --values loadbalancer-tls-disabled.yaml --reuse-values

Reviewer 2: ```suggestion
  helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
    --values loadbalancer-tls-disabled.yaml --reuse-values --wait
```

  ```

  </TabItem>
  <TabItem value="flags" label="--set">

  ```bash

>   helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
>     --set external.enabled=true \
>     --set external.type=LoadBalancer \
>     --set tls.enabled=false

Reviewer 2: ```suggestion
  helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
    --set external.enabled=true \
    --set external.type=LoadBalancer \
    --set tls.enabled=false \
    --wait
```

  ```

  </TabItem>
  </Tabs>

  :::note
  <ValuesYaml path="external"/>
  :::

> 1. Make sure that your managed Kubernetes platform assigned external addresses to your LoadBalancer Services:

Reviewer 2: ```suggestion
1. Make sure that Kubernetes assigned external addresses to your LoadBalancer services:
```

  ```bash
  kubectl get service -n redpanda
  ```

  <details>
  <summary>
  Example output
  </summary>

  ```
  NAME               TYPE           CLUSTER-IP       EXTERNAL-IP      
  lb-redpanda-0      LoadBalancer   10.100.113.102   loadbalancer1.com
  lb-redpanda-1      LoadBalancer   10.100.53.8      loadbalancer2.com
  lb-redpanda-2      LoadBalancer   10.100.231.13    loadbalancer3.com
  ```

  </details>

1. Configure the Redpanda brokers to advertise these external addresses:

  ```bash

>   helm upgrade redpanda redpanda/redpanda -n redpanda --set $(kubectl get svc -n redpanda -o jsonpath='{"external.addresses={"}{ range .items[*]}{.status.loadBalancer.ingress[0].ip }{.status.loadBalancer.ingress[0].hostname}{","}{ end }{"}\n"}') --reuse-values

Reviewer 2: ```suggestion
  helm upgrade redpanda redpanda/redpanda -n redpanda --set $(kubectl get svc -n redpanda -o jsonpath='{"external.addresses={"}{ range .items[*]}{.status.loadBalancer.ingress[0].ip }{.status.loadBalancer.ingress[0].hostname}{","}{ end }{"}\n"}') --reuse-values --wait
```

  ```bash
  kubectl get pod -n redpanda  \
  -o=custom-columns=NODE:.spec.nodeName,NAME:.metadata.name -l \
  app.kubernetes.io/component=redpanda,job-name!=redpanda-redpanda-configuration,job-name!=redpanda-post-upgrade
  ```

1. Make a note of the external addresses of each worker node.

1. Create one load balancer instance outside of your Kubernetes cluster for each worker node that runs Redpanda, and forward the traffic to the node ports that are opened by the LoadBalancer Services.

1. Add the DNS names or IP addresses of your load balancer instances to the `external.addresses` field in order of the StatefulSet replicas. For example, the first address in the list is assigned to `redpanda-0`, the second is assigned to `redpanda-1`, and so on.

  ```bash
  helm upgrade --install redpanda redpanda/redpanda -n redpanda --create-namespace \
    --set "external.addresses={<lb-redpanda-0>,<lb-redpanda-1>,<lb-redpanda-2>}"
    --reuse-values
  ```

1. Use the load balancers' addresses to communicate with the Redpanda cluster from outside the Kubernetes cluster:

  ```bash
  rpk cluster info --brokers <load-balancer-dns-name>:31092
  ```

</TabItem>
</Tabs>

> 

Reviewer 2: Lines 348 to 434 (this line) can be removed, as LoadBalancer services don't require any knowledge or configuration regarding the k8s nodes (so the bare metal and managed k8s sections can be a single section).

Author: @vuldin , I found that if you're not using Managed K8s, such as EC2, you must do these manual steps.

Author: LoadBalancer Services don't get an IP/DNS name in self-managed (bare-metal) environments.

Reviewer 3: I think @vuldin is saying that these steps belong in another place? That maybe these lines are not supposed to be here since they apply to another topic which would be baremetal envs?

Author: These instructions are already within a bare-metal tab. @vuldin do you agree that we should keep these instructions?
